docs/src/CONFIG.md
# Configuring Torrust Tracker
Torrust Tracker's configuration is a simple TOML file. If no TOML file is found, it will run a default configuration.

## Configuration

### Root Level
- `REQUIRED` `mode`: Possible Values: `public`, `listed`, `private` or `private_listed`.
- `REQUIRED` `external_ip`: Set this to your external IP, like `"99.123.43.128"`,
- `OPTIONAL` `log_level`: Possible Values: `off`, `error`, `warning`, `info`, `debug`, `trace`.
- `OPTIONAL` `cleanup_interval`: Interval to clean inactive peers `in seconds`.

### `REQUIRED` `[udp]` Section
- `REQUIRED` `bind_address`: This is where the UDP port will bind to. Example: `0.0.0.0:6969`.
- `REQUIRED` `announce_interval`: Sets the announce interval that will be sent to peers `in seconds`.

### `OPTIONAL` `[http]` Section
- `REQUIRED` `bind_address`: The HTTP REST API will be bound to this address. It's best not to expose this address publicly. Example: `127.0.0.1:80`.

### `REQUIRED IF [http] EXISTS` `[http.access_tokens]` Section
In this section you can make up keys that would be user ids, and values that would be their access token.
If this section is empty, the REST API will not be very useful.

## Sample Configuration
```toml
mode = "public"
external_ip = "0.0.0.0" # set this to your external IP

[udp]
bind_address = "0.0.0.0:6969"
announce_interval = 120 # Two minutes

[http]
bind_address = "127.0.0.1:1212"

[http.access_tokens]
someone = "MyAccessToken"
```


docs/src/config.md
# Configuring UDPT
UDPT's configuration is a simple TOML file.

## Configuration
At the root level, the following options are configurable:
`mode` - Specifies which mode the tracker will operate in. Values can be `static`, `dynamic` or `private`.

### Root Level
- `mode` - Required. Possbile Values: `private`, `static` or `dynamic`.
- `log_level` - Default: `info`. Possible Values: `off`, `error`, `warning`, `info`, `debug`, `trace`.
- `db_path` - Database path. If not set, database will be volatile.
- `cleanup_interval` - Default: 600. Interval to run cleanup in seconds. Cleanup also saves the Database.

### `[udp]` section
This section must exist.

- `bind_address` - Required. This is where the UDP port will bind to. Example: `0.0.0.0:6969`.
- `announce_interval` - Required. Sets the `announce_interval` that will be sent to peers (in seconds).

### `[http]` section
This section is optional.

- `bind_address` - Required (if section exists). The HTTP REST API will be bound to this address. It's best not to expose this address publically. Example: `127.0.0.1:1234`.

### `[http.access_tokens]` section
Section is required if `[http]` section exists.

In this section you can make up keys that would be user ids, and values that would be their access token.
If this section is empty, the REST API will not be very useful.

## Sample Configuration
```toml
mode = "dynamic"
db_path = "database.json.bz2"

[udp]
announce_interval = 120         # Two minutes
bind_address = "0.0.0.0:1212"

[http]
bind_address = "127.0.0.1:1212"

[http.access_tokens]
someone = "MyAccessToken"
```


src/config.rs
pub use crate::tracker::TrackerMode;
use serde::Deserialize;
use std;
use std::collections::HashMap;
use toml;
use std::net::{IpAddr};

#[derive(Deserialize)]
pub struct UDPConfig {
    bind_address: String,
    announce_interval: u32,
}

impl UDPConfig {
    pub fn get_address(&self) -> &str {
        self.bind_address.as_str()
    }

    pub fn get_announce_interval(&self) -> u32 {
        self.announce_interval
    }
}

#[derive(Deserialize)]
pub struct HTTPConfig {
    bind_address: String,
    access_tokens: HashMap<String, String>,
}

impl HTTPConfig {
    pub fn get_address(&self) -> &str {
        self.bind_address.as_str()
    }

    pub fn get_access_tokens(&self) -> &HashMap<String, String> {
        &self.access_tokens
    }
}

#[derive(Deserialize)]
pub struct Configuration {
    mode: TrackerMode,
    udp: UDPConfig,
    http: Option<HTTPConfig>,
    log_level: Option<String>,
    db_path: Option<String>,
    cleanup_interval: Option<u64>,
    external_ip: IpAddr,
}

#[derive(Debug)]
pub enum ConfigError {
    IOError(std::io::Error),
    ParseError(toml::de::Error),
}

impl std::fmt::Display for ConfigError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            ConfigError::IOError(e) => e.fmt(formatter),
            ConfigError::ParseError(e) => e.fmt(formatter),
        }
    }
}
impl std::error::Error for ConfigError {}

impl Configuration {
    pub fn load(data: &[u8]) -> Result<Configuration, toml::de::Error> {
        toml::from_slice(data)
    }

    pub fn load_file(path: &str) -> Result<Configuration, ConfigError> {
        match std::fs::read(path) {
            Err(e) => Err(ConfigError::IOError(e)),
            Ok(data) => {
                match Self::load(data.as_slice()) {
                    Ok(cfg) => {
                        eprintln!("Manually set external IP to: {}", cfg.get_ext_ip());
                        Ok(cfg)
                    },
                    Err(e) => Err(ConfigError::ParseError(e)),
                }
            }
        }
    }

    pub fn get_mode(&self) -> &TrackerMode {
        &self.mode
    }

    pub fn get_udp_config(&self) -> &UDPConfig {
        &self.udp
    }

    pub fn get_log_level(&self) -> &Option<String> {
        &self.log_level
    }

    pub fn get_http_config(&self) -> Option<&HTTPConfig> {
        self.http.as_ref()
    }

    pub fn get_db_path(&self) -> &Option<String> {
        &self.db_path
    }

    pub fn get_cleanup_interval(&self) -> Option<u64> {
        self.cleanup_interval
    }

    pub fn get_ext_ip(&self) -> IpAddr { self.external_ip }
}

impl Configuration {
    pub async fn default() -> Self {
        let external_ip = external_ip::get_ip().await.unwrap();

        eprintln!("external ip: {:?}", external_ip);

        Configuration {
            log_level: Option::from(String::from("info")),
            mode: TrackerMode::PrivateMode,
            udp: UDPConfig {
                announce_interval: 120,
                bind_address: String::from("0.0.0.0:6969"),
            },
            http: Option::from(HTTPConfig {
                bind_address: String::from("127.0.0.1:1212"),
                access_tokens: [(String::from("someone"), String::from("MyAccessToken"))].iter().cloned().collect(),
            }),
            db_path: None,
            cleanup_interval: None,
            external_ip,
        }
    }
}
